WordPress/Classes/Services/PostService.m

```diff
@@ -284,51 +284,71 @@ - (void)uploadPost:(AbstractPost *)post
+        
+        void (^successBlock)(RemotePost *post, NSString *previewURL) = ^(RemotePost *post, NSString *previewURL) {
+            [self.managedObjectContext performBlock:^{
+                AbstractPost *postInContext = (AbstractPost *)[self.managedObjectContext existingObjectWithID:postObjectID error:nil];
+                if (postInContext) {
+                    postInContext.remoteStatus = AbstractPostRemoteStatusAutoSaved;
+                    [self updateMediaForPost:postInContext success:^{
+                        if (success) {
+                            success(postInContext, previewURL);
+                        }
+                    } failure:^(NSError * _Nullable error) {
+                        DDLogInfo(@"Error in updateMediaForPost while remote auto-saving post. description: %@", error.localizedDescription);
+                        // even if media fails to attach we are answering with success because the post auto-save was successful.
+                        if (success) {
+                            success(postInContext, previewURL);
+                        }
+                    }];
+                } else {
+                    // This can happen if the post was deleted right after triggering the auto-save.
                     if (success) {
-                        success(postInContext, previewURL);
+                        success(nil, nil);
                     }
-                }];
-            } else {
-                // This can happen if the post was deleted right after triggering the auto-save.
-                if (success) {
-                    success(nil, nil);
                 }
-            }
-        }];
-    };
-    
-     void (^failureBlock)(NSError *error) = ^(NSError *error) {
-         failure(error);
-     };
-    
-    if ([post.postID longLongValue] > 0) {
-        [remote autoSave:remotePost
-                   success:successBlock
-                   failure:failureBlock];
+            }];
+        };
+        
+        void (^failureBlock)(NSError *error) = ^(NSError *error) {
+            failure(error);
+        };
+        
+        // We should upload the post if:
+        //
+        // - The post is a draft, and has no assigned ID; or
+        // - The post is a draft, and is not accessible through WP.com (because self-hosted doesn't support autosaving).
+        //
+        BOOL needsUploading = [post isDraft] && [post.postID longLongValue] <= 0;
+        
+        if (needsUploading) {
+            [self uploadPost:post
+                     success:^(AbstractPost * _Nonnull post) {
+                         success(post, nil);
+                     }
+                     failure:failure];
+        } else {
+            [restRemote autoSave:remotePost
+                         success:successBlock
+                         failure:failureBlock];
+        }
+    } else if ([post originalIsDraft] && [post isDraft]) {
```

Comment:
As discussed via DM, since we're saving the post in this case, we need to make sure the post was and is still a draft, so that we don't change it's status when we save.

I'll add a comment to document this better.